### PR TITLE
spec.py: stateless satisfies/intersects/constrain

### PR DESCRIPTION
**Pull request by haampie**
Ensure that `Spec.satisfies` and others do not consult the global package repo
to settle whether one is a virtual provided by the other.

In this PR `Spec("mpi").intersects("mpich")` is false as abstract specs.

For concrete specs, the virtuals a node provides according to package.py are frozen on
the node at concretization, with their versions (e.g. `mpi@:3`), so that you can still
do things like:

```
spack find mpi
```

based on "could provide" type of data. When several `provides` clauses for one virtual
match, their versions are intersected, as in the solver.

The data is recorded in the spec file as `provided_virtuals` and is part of the DAG
hash, so this changes hashes. That requires a format bump:

* spec file v6, database v9, lockfile v8
* buildcache media types `spec.v6` and `db.v9`

For older versions, this data is reconstructed upon read. The bootstrap prototypes
are upgraded once and for all (cheating a little bit about the DAG hash, but not a
big deal).

This PR also reverts #40898, which was to avoid hitting the slow path of `satisfies(x, y)` 
that triggered the provider index being created or being validated, which is at a minimum
9k stat calls. That serves as a motivating example of why this PR is desirable -- there are
many more cases where a satisfies call can be unexpectedly slow on develop.

### CODE REVIEW

File: lib/spack/spack/schema/spec.py [hunk 1]
```diff
@@ -235,6 +235,12 @@
             "items": {"type": "string"},
             "description": "List of patches, similar to the patches variant under parameters",
         },
+        "provided_virtuals": {
+            "type": "object",
+            "additionalProperties": {"type": "string"},
+            "description": "Virtual name -> provided version(s), frozen at concretization time. "
+            "Part of the DAG hash, since `provides` is stripped from the package hash.",
+        },
```
Comment (alalazo) [inline lib/spack/spack/schema/spec.py:243]: Why a map rather than a list of strings
```
provided_virtuals: ["mpi@3:", "lapack"]
```
? Does it make any performance difference?

Comment (alalazo) [reply lib/spack/spack/schema/spec.py:243]: Also: if we do this should we also think forward about virtuals that may have variants, and add both the schema and the capability to the reader? Would that help us in the future to avoid another bump in the spec format?

File: lib/spack/spack/repo.py [hunk 2]
```diff
@@ -2156,6 +2159,110 @@ def create_and_enable(
...
+
+        for spec in nodes:
+            if spec.concrete and spec._provided_virtuals is None:
+                spec._provided_virtuals = _provided_versions(spec, providers.get(spec.name, ()))
+
+    for spec in nodes:
+        if not spec.concrete:
+            continue
+
+        if not edges_lack_virtuals and spec.original_spec_format() >= 3:
```
Comment (alalazo) [inline lib/spack/spack/repo.py:2240]: (Possibly a bug?)
```suggestion
        if not edges_lack_virtuals and spec.original_spec_format() >= 4:
```
? I think currently we skip reconstruction on Specfile v3?

**Review by alalazo — DISMISSED** [review]
As discussed in DM this should come with a Specfile version bump (which then cascades to a a DB bump and a lockfile bump). 

The PR should likely be labeled as `hash-change` too?

**Comment by alalazo** [comment]
This is half-joking, but this kind of change would better be captured by a minor version bump in the specfile - should we discuss that?

File: lib/spack/spack/schema/spec.py (hunk 1 above)
Comment (haampie) [reply lib/spack/spack/schema/spec.py:255]: Looks better to me as well.

**Comment by haampie** [comment]
> This is half-joking, but this kind of change would better be captured by a minor version bump in the specfile - should we discuss that?

Feel free, but that only helps for the first forward incompatible change after Spack v1.3, so not relevant to this PR.

File: lib/spack/spack/repo.py (hunk 2 above)
Comment (haampie) [reply lib/spack/spack/repo.py]: Fixed

**Comment by alalazo** [comment]
I'm currently torn on this one (I'd like to merge it soon, but I'd also like to signal in some way that we added more information in the spec file).

A few thoughts in random order:
1. Bumping the specfile version is a major usability issue, and this is not a fault of the PR
2. We could go back to avoid requiring bump for this PR. In particular if we spell out `provided_virtuals: []` there's no double meaning on what the absence of that key means.
3. Since we reconstruct from the current repo we may end up with cases where a node provides no virtuals, but it has virtuals on incoming edges (e.g. a directive was removed in the meanwhile). I guess this is very rare, and only occurs when reconstructing old specfiles. For new spec files everything is consistent as it comes from the solver.
4. This PR changes the hashes of all spec concretization due to the new field, so the same freshly concretized spec with the same package hash will change across the PR boundary. Is that enough to call for a bump here? Maybe not?

File: lib/spack/spack/spec.py [hunk 3]
```diff
@@ -2721,6 +2700,13 @@ def to_node_dict(self) -> Dict[str, Any]:
         if self._package_hash:
             d["package_hash"] = self._package_hash
 
+        # Since v6, an absent key on a concrete node means it provides nothing
+        if self._concrete and self.provided_virtuals:
```
Comment (alalazo) [inline lib/spack/spack/spec.py:2704]: The `to_node_dict` complexity rises in this PR. Can we extract some helper function for `provided_virtuals`?

File: lib/spack/spack/repo.py [hunk 4]
```diff
@@ -2156,6 +2159,111 @@ def create_and_enable(
...
+    for spec in nodes:
+        if not spec.concrete:
+            continue
+
+        # Edge virtuals are recorded from spec format v4 on
+        if not edges_lack_virtuals and spec.original_spec_format() >= 4:
+            continue
+
+        try:
+            pkg_cls = PATH.get_pkg_class(spec.fullname)
```
Comment (alalazo) [inline lib/spack/spack/repo.py:2245]: Can we inject the repository here?

File: lib/spack/spack/spec.py [hunk 5]
```diff
@@ -3800,6 +3659,8 @@ def _dup(
                 If deptype, or depflag, copy matching types.
         """
         self._package = None
+        # Shared, not copied: the map is only ever replaced as a whole, never mutated in place.
+        self._provided_virtuals = other._provided_virtuals
```
Comment (alalazo) [inline lib/spack/spack/spec.py:3663]: (question) Do we have ways to enforce this?

**Review by alalazo — DISMISSED** [review]
On the merge branch with `develop`:
```console
$ spack spec --json --fresh gcc@14 languages=c,c++,go
```
produces the following specfile that cannot be read back:
```json
...
        "provided_virtuals": [
          "c",
          "cxx",
          "go-or-gccgo-bootstrap@", 
          "golang@"
        ],
```
Note the `@` without a following version constraint. We should fix this before merging. (This comes from directives that are written in the wrong way, but we shouldn't emit unparseable strings)

**Comment by haampie** [comment]
Kinda concerning that we "independently" find the same edge cases that would have gone unnoticed for I don't know how long. Does that mean that Claude is really good at finding that particular edge case, or all edge cases? :man_shrugging:

File: lib/spack/spack/spec.py (hunk 5 above)
Comment (haampie) [reply lib/spack/spack/spec.py:3663]: It was a map, it's a list now in other code paths, isn't it? I can make it a tuple if you care, at the cost of a linear scan.

File: lib/spack/spack/test/spec_semantics.py [hunk 6]
```diff
@@ -856,6 +858,131 @@ def test_concrete_checks_on_virtual_names_dont_need_repo(self, monkeypatch):
 
         assert not concrete.satisfies("%c,mpi=mpich")
 
+    @staticmethod
+    def _old_spec_dict(spec: Spec) -> dict:
```
Comment (alalazo) [inline lib/spack/spack/test/spec_semantics.py:861]: I think we should also set `original_specfile_version` to `5` to mock an old spec.

File: lib/spack/spack/test/spec_semantics.py [hunk 7]
```diff
@@ -856,6 +858,131 @@ def test_concrete_checks_on_virtual_names_dont_need_repo(self, monkeypatch):
...
+    def test_old_specfile_with_unknown_package_reads_without_warning(self):
+        """A package absent from the configured repos provides nothing, without warnings."""
+        as_dict = self._old_spec_dict(spack.concretize.concretize_one("pkg-a"))
+        as_dict["spec"]["nodes"][0]["name"] = "no-such-package"
+
+        with warnings.catch_warnings():
+            warnings.simplefilter("error")
+            reread = Spec.from_dict(as_dict)
+
+        assert reread.provided_virtuals == ()
```
Comment (alalazo) [inline lib/spack/spack/test/spec_semantics.py]: Why do we test specifically that we don't warn in this case?

File: lib/spack/spack/database.py [hunk 8]
```diff
@@ -923,6 +925,10 @@ def invalid_record(hash_key, error):
         for hash_key, rec in data.items():
             rec.spec._mark_root_concrete()
 
+        # Pass 4: reconstruct the virtual data that spec formats before v6 omit
+        if spec_reader.SPEC_VERSION < 6:
+            spack.repo.reconstruct_virtuals([rec.spec for rec in data.values()])
```
Comment (alalazo) [inline lib/spack/spack/database.py]: I think removing these two lines leaves all the tests passing. Can we add a test that checks this part?

File: lib/spack/spack/environment/environment.py [hunk 9]
```diff
@@ -2541,6 +2541,10 @@ def _filter_specs(self, reader, json_specs_by_hash, order_concretized):
                 _, bhash, _ = reader.extract_build_spec_info_from_node_dict(node_dict)
                 specs_by_hash[lockfile_key]._build_spec = specs_by_hash[bhash]
 
+        # The DAG is wired by hand above, so reconstruct what spec formats before v6 omit
+        if reader.SPEC_VERSION < 6:
+            spack.repo.reconstruct_virtuals(specs_by_hash.values())
```
Comment (alalazo) [inline lib/spack/spack/environment/environment.py]: Same for this one.

**Review by alalazo — COMMENTED** [review]
I think at this point the remaining changes are minor or nits:
1. `reconstruct_virtuals` could take a `repo=` argument, and callers can explicitly pass `spack.repo.PATH` or whatever repository they have in scope.
2. Add unit tests to prevent regressions in reconstruction (see comments)

There is also an edge case that can create a mismatch between the `provided_virtuals` and the `virtuals_on_edges`. If an old specfile:
1. Has a virtual on an edge, and
2. In the meanwhile the `provides` directive gets deleted in the package recipe

We may end up with an inconsistency (i.e. no provided virtual from a node that has that virtual on an edge). I'm honestly thinking of accepting this as a known shortcoming of the reconstruction, considering how rare it would be to hit it.

---

OT: My understanding of the discussion last week is that after this PR bumps the Specfile version to v6 etc. all the other PRs slated for Spack v1.3 that also affect the Specfile will:
-  add to the Specfile v6 changes, but
-  won't bump the version

Repeating it here to make sure we all understood the same and we are aligned on that 🙂

File: lib/spack/spack/url_buildcache.py [hunk 10]
```diff
@@ -144,17 +154,23 @@ def from_dict(cls, manifest_json: Dict[str, Any]) -> "BuildcacheManifest":
...
+        matches = [r for r in self.data if r.media_type == media_type]
+        wanted = _SPACK_MEDIA_TYPE.fullmatch(media_type)
+        if not matches and wanted:
+            for record in self.data:
+                found = _SPACK_MEDIA_TYPE.fullmatch(record.media_type)
+                if (
+                    found
+                    and found["family"] == wanted["family"]
+                    and found["suffix"] == wanted["suffix"]
+                    and int(found["version"]) < int(wanted["version"])
```
Comment (kwryankrattiger) [inline lib/spack/spack/url_buildcache.py]:
```suggestion
                    and int(found["version"]) <= int(wanted["version"])
```

Comment (kwryankrattiger) [reply lib/spack/spack/url_buildcache.py]: This is also not quite right, there are a number of versions we don't expect at all for the `v3` cache layout.

Comment (haampie) [reply lib/spack/spack/url_buildcache.py]: now it looks more like your pr

**Review by alalazo — COMMENTED** [review]
I think Claude spotted another subtle bug: provided virtuals are frozen too late. 

The `finalize_concretization` function freezes them:
1. after computing package hashes, and 
2. after `_inject_patches_variant` runs 

This means `_provided_virtuals` is still `None` on every node in both cases above. So `_provides_virtual` returns `False`, and any condition on a *versioned* virtual evaluates to false.[^1]

[^1]: Conditions on an unversioned virtual like `^mpi` still work, because `_satisfies_edge_attributes` returns early on `any_version`.

Note that this is not a latent bug, but a real one. `hdf` has:
```python
patch("disable_doclint.patch", when="@:4.2.14^java@9:")
```

With `spack spec --fresh --json 'hdf@4.2.14+java+shared ^openjdk@11'`:

- on `develop`: `hdf` has `patches: ['80b233283a346f7d9ad3cd9a7360c100055b84fe29840d705617ffeb81f56298']`
- on this branch: `hdf` has no `patches` at all, although `openjdk` correctly ends up with `provided_virtuals: ['java@11']`

**Comment by haampie** [comment]
I'll leave this comment for @becker33 to think about, cause I don't consider it a blocker for this PR: if you mutate the versions of a concrete spec, you could argue it invalidates the "could provide" data. At the same time, the user did not mutate the "could provide" data, and the user can already make modifications to concrete specs that are invalid regarding `package.py` constraints. So, it's unclear whether mutation should be followed by updating `"provided_virtuals"`. Similarly, unclear if the package hash should be reassigned after mutation, patches whose when condition now trigger should be set, ...

**Comment by haampie** [comment]
Another nit is that the solver does not consider the `"provided_virtuals"` data, and instead uses the (modified) `package.py` as its source of truth. I think that's fine, cause the spec attribute is opaque, whereas the package.py file is easy to inspect.

**Review by alalazo — APPROVED** [review]
PR looks good to me. One last caveat is that currently:
```console
$ spack buildcache push --update-index
```
on a shared layout-v3 mirror becomes unusable for Spack v1.2 and earlier. That can be fixed in a separate PR, by e.g. having separate indexes.

**Comment by haampie** [comment]
Let's see how much work that is, cause if we merge this and someone happens to have `spack buildcache push --update-index` in their CI with Spack `develop`, it'd be annoying to break forward compat temporarily, and they would have to reindex with old Spack before reindexing with even newer Spack.

**Comment by haampie** [comment]
https://github.com/spack/spack/pull/53094 is simple enough to go first.

**Review by alalazo — DISMISSED** [review]
#53094 is in I think we're ready to trigger the specfile version bump 🤞

File: lib/spack/spack/test/binary_distribution.py [hunk 11]
```diff
@@ -1287,7 +1289,7 @@ def mock_index(tmp_path: pathlib.Path, monkeypatch) -> IndexInformation:
 
     index_blob_record = spack.binary_distribution.BlobRecord(
         os.stat(index_blob_path).st_size,
-        cache_class.BUILDCACHE_INDEX_MEDIATYPE,
+        cache_class.component_to_media_type(BuildcacheComponent.INDEX),
```
Comment (kwryankrattiger) [inline lib/spack/spack/test/binary_distribution.py]:
```suggestion
        cache_class.current_to_media_type(BuildcacheComponent.INDEX),
```

File: lib/spack/spack/test/binary_distribution.py [hunk 12]
```diff
@@ -1827,3 +1829,21 @@ def __str__(self):
...
+@pytest.mark.parametrize(
+    "component,oldest",
+    [
+        (BuildcacheComponent.SPEC, "application/vnd.spack.spec.v5+json"),
+        (BuildcacheComponent.INDEX, "application/vnd.spack.db.v8+json"),
+    ],
+)
+def test_manifest_reads_older_media_types(component, oldest):
+    """Blobs of the formats layout v3 started with are still read, after the current ones."""
+    current = URLBuildcacheEntry.component_to_media_type(component)
```
Comment (kwryankrattiger) [inline lib/spack/spack/test/binary_distribution.py]:
```suggestion
    current = URLBuildcacheEntry.current_to_media_type(component)[0]
```

**Review by kwryankrattiger — CHANGES_REQUESTED** [review]

File: lib/spack/spack/buildcache_migrate.py [hunk 13]
```diff
@@ -186,12 +186,17 @@ def _migrate_spec(
     )
 
     tarball_blob_record = BlobRecord(
-        spec_dict["archive_size"], v3_cache_class.TARBALL_MEDIATYPE, "gzip", algorithm, checksum
+        spec_dict["archive_size"],
+        v3_cache_class.component_to_media_type(BuildcacheComponent.TARBALL),
```
Comment (kwryankrattiger) [inline lib/spack/spack/buildcache_migrate.py]:
```suggestion
        v3_cache_class.component_to_media_types(BuildcacheComponent.TARBALL)[-1],
```

Since this a a v2 -> v3 migration, that implies that the tarball is the oldest supported composition.

**Comment by alalazo** [comment]
@kwryankrattiger Can you explain the blocker? I don't get why you suggest to take a char out of a string? Note the `*_media_type ->str` vs. `*_media_types -> List[str]`.

File: lib/spack/spack/url_buildcache.py [hunk 14]
```diff
@@ -323,21 +329,18 @@ def get_buildcache_component_include_pattern(
...
-            return cls.PUBLIC_KEY_INDEX_MEDIATYPE
+    def component_to_media_types(cls, component: BuildcacheComponent) -> List[str]:
+        """Media types of a buildcache component that are read, newest first"""
+        if component in cls.MEDIA_TYPES:
+            return cls.MEDIA_TYPES[component]
 
         raise BuildcacheEntryError(f"Not a blob component: {component}")
 
+    @classmethod
+    def component_to_media_type(cls, component: BuildcacheComponent) -> str:
```
Comment (kwryankrattiger) [inline lib/spack/spack/url_buildcache.py]: idk if I like this API, it seems to have already confused me in this review!


```suggestion
    def current_component_to_media_type(cls, component: BuildcacheComponent) -> str:
```

**Comment by kwryankrattiger** [comment]
@alalazo 
> Can you explain the blocker?

Yeah, so I was part way through the review and ran into those and submitted it quickly to put the block in while I finished the review. The API is the real problem, I updated the suggestions to reflect my suggestion. If I am making this mistake in the review it means it is too hard to distinguish the APIs imo, the the name should be sufficiently different to avoid confusion. As a rule of thumb I do not like the usage of plural to distinguish APIs, it is too easy to confuse people and is also often not specific enough to imply intent. It isn't just giving a single type, it is giving the current type supported by Spack for pushing new things.

File: lib/spack/spack/buildcache_migrate.py (hunk 13 above)
Comment (kwryankrattiger) [reply lib/spack/spack/buildcache_migrate.py]: This could also have a specialized API call, we do the same below.

```suggestion
        v3_cache_class.oldest_component_to_media_type(BuildcacheComponent.TARBALL),
```

**Review by kwryankrattiger — DISMISSED** [review]

**Review by alalazo — APPROVED** [review]
